internal/arp/arp_test.go

```diff
@@ -0,0 +1,169 @@
+			},
+			reason: dropReasonEthernetDestination,
+		},
+		{
+			name:   "not announcing IP",
+			modify: func(_ *ethernet.Frame, _ *arp.Packet) {},
+			reason: dropReasonAnnounceIP,
+		},
+		{
+			name: "not leader",
+			modify: func(_ *ethernet.Frame, p *arp.Packet) {
+				p.TargetIP = announceIP
+			},
+			announce: announceIP,
+			reason:   dropReasonNotLeader,
+		},
+		{
+			name: "OK",
+			modify: func(_ *ethernet.Frame, p *arp.Packet) {
+				p.TargetIP = announceIP
+			},
+			announce: announceIP,
+			leader:   true,
+			reason:   dropReasonNone,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			a, conn, done := newAnnounce(t)
+			defer done()
+
+			// Set leader status and announce IP if necessary.
+			a.leader = tt.leader
+
+			if tt.announce != nil {
+				a.SetBalancer("test", announceIP)
+			}
+
+			// Stock data which can be overridden by modify function.
+			var (
+				srcHW = net.HardwareAddr{0xde, 0xad, 0xbe, 0xef, 0xde, 0xad}
+				srcIP = net.IPv4(192, 168, 1, 1).To4()
+
+				dstHW = a.hardwareAddr
+				dstIP = net.IPv4(192, 168, 1, 10).To4()
+			)
+
+			eth := &ethernet.Frame{
+				Destination: a.hardwareAddr,
+				Source:      srcHW,
+				EtherType:   ethernet.EtherTypeARP,
+			}
+
+			pkt, err := arp.NewPacket(arp.OperationRequest, srcHW, srcIP, dstHW, dstIP)
+			if err != nil {
+				t.Fatalf("failed to make ARP packet: %v", err)
+			}
+
+			// Modify the frame and packet before marshaling.
+			tt.modify(eth, pkt)
+
+			eth.Payload = mustMarshal(pkt)
+			b := mustMarshal(eth)
+
+			// Wait for a single packet to be received.
+			var wg sync.WaitGroup
+			wg.Add(1)
+			defer wg.Wait()
+
+			dropC := make(chan dropReason)
+			go func() {
+				defer wg.Done()
+
+				dropC <- a.readPacket()
+			}()
+
+			// Send a packet to receiver goroutine.
+			if _, err := conn.Write(b); err != nil {
+				t.Fatalf("failed to write: %v", err)
+			}
+
+			reason := <-dropC
+			if diff := cmp.Diff(tt.reason, reason); diff != "" {
+				t.Fatalf("unexpected drop reason (-want +got)\n%s", diff)
+			}
+		})
+	}
+}
+
+func mustMarshal(m encoding.BinaryMarshaler) []byte {
+	b, err := m.MarshalBinary()
+	if err != nil {
+		panic(fmt.Sprintf("failed to marshal: %v", err))
+	}
+
+	return b
+}
+
+func newAnnounce(t *testing.T) (*Announce, *net.UDPConn, func()) {
+	pc, err := net.ListenPacket("udp", ":0")
+	if err != nil {
+		t.Fatalf("failed to listen UDP: %v", err)
+	}
+
+	// TODO(mdlayher): don't depend on an interface when we really just care
+	// about its addresses anyway.
+	ifi, err := net.InterfaceByName("eth0")
+	if err != nil {
+		t.Skipf("skipping, couldn't get eth0: %v", err)
+	}
+
+	c, err := arp.New(ifi, pc)
+	if err != nil {
+		t.Fatalf("failed to create ARP client: %v", err)
+	}
+
+	a := &Announce{
+		hardwareAddr: ifi.HardwareAddr,
+		client:       c,
+		ips:          make(map[string]net.IP),
+	}
+
+	uc, err := net.DialUDP("udp", nil, pc.LocalAddr().(*net.UDPAddr))
```

Comment:
This fails on travis - looks like the packetconn sets the IP portion of the address to the empty address (`[::]`), so the dial fails?

Seems to fail consistently on 1.8, 1.9, 1.10beta1. Can you reproduce locally?

Reply:
Interesting, it works just fine locally with 1.9.2 but I'll play with it more tomorrow.

Reply:
doesn't get `:0` you an IP v6 addr? that fails on travis?

Reply:
Updated to use `localhost:0` instead and it works in Travis now.

Reply:
Turns out this is travis-ci/travis-ci#5359 . IPv6 connectivity in test containers used to work, then they deliberately broke it because they don't *officially* support it :(

I guess I need to think about where to rehome the CI/CD stuff some day...